src/service/backup_service.rs

```diff
@@ -0,0 +1,135 @@
+use std::sync::Arc;
+
+use crate::{
+    persistence::{backup::BackupStoreApi, db::SurrealDbConfig, identity::IdentityStoreApi},
+    util,
+};
+
+use super::{Error, Result};
+#[cfg(test)]
+use mockall::automock;
+
+/// Allows to backup and restore the database as an encrypted file.
+#[cfg_attr(test, automock)]
+#[async_trait::async_trait]
+pub trait BackupServiceApi: Send + Sync {
+    /// Creates an encrypted backup of the database and returns the
+    /// download path to the backup file.
+    async fn backup(&self) -> Result<Vec<u8>>;
+}
+
+pub struct BackupService {
+    store: Arc<dyn BackupStoreApi>,
+    identity_store: Arc<dyn IdentityStoreApi>,
+    surreal_db_config: SurrealDbConfig,
+}
+
+impl BackupService {
+    pub fn new(
+        store: Arc<dyn BackupStoreApi>,
+        identity_store: Arc<dyn IdentityStoreApi>,
+        surreal_db_config: SurrealDbConfig,
+    ) -> Self {
+        Self {
+            store,
+            identity_store,
+            surreal_db_config,
+        }
+    }
+
+    fn validate_surreal_db_connection(&self) -> Result<()> {
+        let connection = &self.surreal_db_config.connection_string;
+        if connection.starts_with("rocksdb")
+            || connection.starts_with("mem")
+            || connection.starts_with("http")
+        {
+            return Ok(());
+        }
+        Err(Error::Validation(format!(
+            "SurrealDB connection {} does not support exports",
+            connection
+        )))
+    }
+}
+
+#[async_trait::async_trait]
+impl BackupServiceApi for BackupService {
+    async fn backup(&self) -> Result<Vec<u8>> {
+        self.validate_surreal_db_connection()?;
+        let public_key = self
+            .identity_store
+            .get_full()
```

Comment:
not too relevant, but you could call `get_key_pair` here if you only need the key